### ISSUE TITLE
ID tool pulling P corni when no results should appear

### ISSUE THREAD

**Issue by Megachile** [jeffdc/gallformers#257]
No Quercus species are currently mapped to Alberta. However, selecting both quercus and alberta in the ID tool incorrectly yields one result: P corni. 
![alberta](https://user-images.githubusercontent.com/69063635/132140494-f0952503-4450-4f5f-ae6a-0808515afd38.JPG)

**Comment by jeffdc** [comment]
Poking at this a bit it seems that if you select any combination of genus and place that should yield zero results it returns P corni. Very strange. I will look at it.

**Comment by jeffdc** [comment]
Here is what is happening:

1. A genus is selected, say 'Nyssa'
2. All galls for Nyssa are fetched from the back end
3. A Place filter is applied, say 'Alberta'
4. The set of galls (already fetched from the back end) is filtered in the browser based on the list of Places that the Gall occurs in (which was formed in step 2 by looking at all of the hosts for the gall and pulling all of the places and and aggregating into a single list)
5. P corni will appear even though Nyssa does not occur in Alberta

P corni shows up because it occurs on Nyssa AND it occurs on other hosts, e.g. Acer, that do appear in Alberta. Since the list of Galls and the Places was generated by fetching data from the DB it is not possible to filter P corni out for this combination as the Host-Place data is not available (it is not passed down from the database as to do so would make the ID tool very slow since so much data would have to be passed).

This can be fixed but it will require a re-think of how the ID page works under the hood. 

Options:

1. Have the Place field, when changed, re-fetch data from the back end. 
2. Pass the host genus info down with the results so that it can be filtered correctly on the client side
3. Re-implement the page to not pre-fetch ANY data and instead have it all happen in real-time

Options 1 and 2 are much easier than 3. Option 2 may have performance impacts since we will have to pass down a lot more data (every gall returned will have to have a list of Hosts attached to them). Option 1 will require a slight re-think of the UI since the Place field will behave more like the Host/Genus(Section) fields then any of the other filter fields.

As I have mentioned in the past, I think that the ID tool needs a total re-think. It is at the heart of the User's experience when using the site and I built it with only a partial understanding of how it might be used (as all software is built!). Now that I have a better handle on it, using it a lot myself, I have a lot of ideas on how to make it much better. The problem is that it is a lot of work to do this and I have limited time and we have no one else doing any tech work.

I think that for now we should go with Option 1 above. I may run into other issues with this approach because of assumptions about the clear separation between selecting a Host and all of the Filter properties that currently exists.

**Comment by Megachile** [comment]
> P corni shows up because it occurs on Nyssa AND it occurs on other hosts, e.g. Acer, that do appear in Alberta

I think I get what you're saying, but if so, it seems like this would cause issues more frequently than I've noticed. Maybe it has fetched "wrong" answers I haven't noticed at other times, but I guess if it's always within the genus then they wouldn't strike me as odd. Anyway, whichever solution you decide makes sense works for me--or honestly, if you want to wait and just leave this issue until you redo the ID page in general I don't think that's a huge problem.

**Comment by jeffdc** [comment]
I think that since most galls do not cross genera this issue does not show up frequently. e.g.,

https://www.gallformers.org/id?hostOrTaxon=Cornus&type=genus&detachable=&alignment=&cells=&color=&locations=&season=&shape=&textures=&walls=&form=&undescribed=false&place=Quebec

In addition to P coni, Pulvinaria acericola and Neopulvinaria innumerabilis shoudl not be here. They both occur on Cornus florida but C florida does not occur on Quebec.

**Comment by Megachile** [comment]
Yes, though those P corni examples are probably a case where the mistake corrects for an inappropriate narrowness of our data. It's likely that P corni is found on Cornus in Quebec even if it's not P corni.
